stm32: Add reinitialization feature for SPI and UART

knthm opened the pull request:
This PR adds device reinitialization for SPI and UART, as these will lose their hardware configuration when certain SoC series enter certain STOP/Standby modes.

Unfortunately, I've had to remove ADC and I2C reinit from this PR, as they require deeper PM work, seeing how they need to block during reinit process.

#### Depends on commits from PRs

- #60998 
- #60369

#### References

See RM0461, chapter 5.4 - Low-Power Modes, Table 37, and chapter 5.4.9 - Stop 2 Mode:

1. [RM0461 Reference manual - STM32WLEx advanced Arm-based 32-bit MCUs with sub-GHz radio solution](https://www.st.com/resource/en/reference_manual/rm0461-stm32wlex-advanced-armbased-32bit-mcus-with-subghz-radio-solution-stmicroelectronics.pdf)

knthm committed:
pm: Add reinit-power-states DT binding property

This property allows defining one or more power-states phandles in which
a peripheral loses its configuration, and must be reinitialized.

States below and including S2RAM are inherently assumed to require
reinit, this property is intended to cover the exceptions outside
of this common use case.

It is intended to assist drivers in preventing the SoC from entering
these power states during critical code sections, as well as indicate
to power management services when a peripheral needs reinitialization.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
dts: arm: st: wl: Enable reinit for select peripherals

Enable STOP2 reinitialization for affected USART and SPI peripheral
instances, as these will otherwise be INOP when the SoC returns from
STOP2 low-power mode.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
drivers: spi: stm32: Refactor for reinitialization feature

Refactor to prevent code duplication in the driver reinitialization
functionality following this commit.

Move HAL include statement from driver source to header to make header
self-contained.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

erwango commented on drivers/serial/uart_stm32.c:
```diff
@@ -1628,11 +1614,15 @@ static int uart_stm32_init(const struct device *dev)
+	return 0;
+}
+
+static int uart_stm32_registers_configure(const struct device *dev)
```
Currently `uart_stm32_configure` doesn't save any application configuration in instance data.
Hence, if application performed a runtime configuration before a power state where registers are reset, this configuration will be lost so need to be performed again.
Do you think this can be performed by application ? (which implies that application needs to keep track of device state and needs)
If runtime configuration restoration has to be performed by device itself, it needs to be able to store it and re-apply when needed.

Applicability of this point depends on runtime configuration potential of the device API in use.

knthm replied:
That's a very good point, I hadn't considered this.
My use case for the UART changes was to ensure that the serial console output worked when the SoC comes out of STOP2, in order to assist debugging. :smile:
AFAICT, general usage of `CONFIG_UART_CONSOLE` doesn't make use of the `uart_configure` syscall, which would trigger `uart_stm32_configure`, but please correct me if I'm wrong.

I don't think the application should keep track of, or be required to restore the UART runtime configuration, that feels more like the scope of the OS services.

My first go at this would be to add a struct to hold the runtime configuration, and if that's been initialized by an application calling `uart_configure` at some point, the UART reinit function would prefer restoring the parameters stored there rather than the generic register initialization also done on boot.

There's significant overlap between the newly added `uart_stm32_registers_configure` functionality, originally taken from `uart_stm32_init`, and `uart_stm32_configure`, so I'll try to unify the behavior of these with the new struct in a way that's more effective and less redudant than the current code.

erwango replied:
> My first go at this would be to add a struct to hold the runtime configuration, and if that's been initialized by an application calling uart_configure at some point, the UART reinit function would prefer restoring the parameters stored there rather than the generic register initialization also done on boot.

Sounds reaonable

> There's significant overlap between the newly added uart_stm32_registers_configure functionality, originally taken from uart_stm32_init, and uart_stm32_configure, so I'll try to unify the behavior of these with the new struct in a way that's more effective and less redudant than the current code.

Good point. Don't hesitate to split this work in a different PR to focus current PR on its initial topic.

knthm replied:
> Good point. Don't hesitate to split this work in a different PR to focus current PR on its initial topic.

Since the changes are tightly coupled with the reinit logic used here, I've decided to amend the commit with my above suggestion. If you think it muddies the scope of this PR too much, and you'd prefer I put it in a separate PR, just let me know and I'll oblige.

The commit message describes the changes pretty well I think, but here's some key points to aid review:

- Added static `uart_config` struct to `STM32_UART_INIT` macro
- Added `uart_config` struct pointer as `uart_stm32_config` member
- (Re)moved now superfluous `baud_rate`, `parity`, and `hw_flow_control` struct members
- Moved overlapping baudrate/parity/data-/stop-bits/flow control configuration to common function, with persistence in static instance `uart_config` struct
- Removed incorrect runtime parity/data bits configuration, and replaced with thoroughly tested compile-time `BUILD_ASSERT` statements
  - Original runtime logic in `init` forced a very small subset of UART configurations, and straight-up ignored many valid DT property combinations (e.g 7E1)
  - This now also allows invalid configurations to get flagged compile-time
  - The previous logic partially conflicted with the logic in `uart_stm32_configure`, especially that relating to parity/data-bits combinations, either enforced opposite configurations
- The `#if defined(LL_USART_STOPBITS_1_5) && HAS_LPUART_1` conditional broke "half-bit" parity application configurations of USARTs, this is now fixed

Overall the driver will now behave more coherently, regardless if configured by DTS on (re-)init, or by application/subsystem syscall.

knthm replied:
Some follow up questions on things that don't make sense to me, and might need attention as well:

- Is there any reason the driver forbids runtime configuration with [9 data bits](https://github.com/zephyrproject-rtos/zephyr/blob/89dabaeec2c2485c0a256ee6465dd5c0878a4fc8/drivers/serial/uart_stm32.c#L523-L524)? A quick grep of the HAL module shows all STM32 series UART support 9 data bits.
- Why does the runtime configuration [default to 2 stop bits](https://github.com/zephyrproject-rtos/zephyr/blob/89dabaeec2c2485c0a256ee6465dd5c0878a4fc8/drivers/serial/uart_stm32.c#L320-L321) instead of 1?

erwango replied:
> Some follow up questions on things that don't make sense to me, and might need attention as well:

> Is there any reason the driver forbids runtime configuration with [9 data bits](https://github.com/zephyrproject-rtos/zephyr/blob/89dabaeec2c2485c0a256ee6465dd5c0878a4fc8/drivers/serial/uart_stm32.c#L523-L524)? A quick grep of the HAL module shows all STM32 series UART support 9 data bits.
Why does the runtime configuration [default to 2 stop bits](https://github.com/zephyrproject-rtos/zephyr/blob/89dabaeec2c2485c0a256ee6465dd5c0878a4fc8/drivers/serial/uart_stm32.c#L320-L321) instead of 1?

This might actually be a genuine bug. See related https://github.com/zephyrproject-rtos/zephyr/issues/56003 for instance

erwango replied:
There's certainly a good effort to document the change on this commit, but I'd prefer this is split in a dedicated PR (and potentially few commits if doable).
This will definitely ease and speed up the review (avoid mixing topics, ..).
Please also note related change https://github.com/zephyrproject-rtos/zephyr/pull/54354 that might get across your changes

knthm replied:
Sure thing, I'll move all UART changes into a separate PR when I find a spare few minutes today. 

That means this PR won't touch the UART driver until it's merged. If this should get merged first, I'll have to add the UART reinit calling via `pm_action` separately–but that should be very similar to the other drivers', based on my proposed implementation of this mentioned below.

erwango commented on dts/bindings/i2c/st,stm32-i2c-v2.yaml:
```diff
@@ -55,3 +55,9 @@ properties:
       GPIO to which the I2C SDA signal is routed. This is only needed for
       I2C bus recovery support.
+
+  reinit-power-states:
+    type: phandles
+    description: |
+      Power states where the peripheral context is lost, and must be
+      reinitialized to ensure correct operation.
```
Idealy, this should stand in `dts/bindings/base/pm.yaml`. @ceolin to confirm

erwango commented on soc/arm/st_stm32/stm32wl/power.c:
```diff
@@ -76,20 +86,34 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+		case 2: /* STOP1 */
+			break;
+		case 3: /* STOP2 */
+			/* Reinitialize peripheral functionality */
+			DT_FOREACH_STATUS_OKAY(st_stm32_adc,    STM32_ADC_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_i2c_v2, STM32_I2C_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_spi,    STM32_SPI_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_usart,  STM32_USART_REINIT);
```
This is IMO the most controversial part of this work:
- This is performed before `stm32_clock_control_init()` and potentially this could have some impacts in some clock configuration for some peripherals
- This does not take into account actions performed in device's `_pm_action()` function.
- This is not ideal to maintain
- Ideally, there should be a way, in device's `_pm_action()` functions to take into account the PM state we're coming back and act accordingly. Maybe @ceolin can comment on this assertion.

knthm replied:
Good points, let me try to devise a better approach.

knthm replied:
After some deeper review of the Device PM documentation and code, I've realized I had some false presumptions about how it worked in relation to System PM. I'd initially thought they worked somewhat independently, but now I see that System PM state changes will trigger Device PM actions.

I think the following would be the cleanest implementation of the reinit logic with Device PM:
1. In the `pm.c` `pm_suspend_devices` and `pm_resume_devices` functions, a `pm_state_info` struct pointer (from `z_cpus_pm_forced_state[CURRENT_CPU]`) is passed to the individual drivers' `*_pm_action` callback, as a (new) third parameter.
2. The drivers will then have access to the System PM power-state that triggered their `*_pm_action` callback.
From there they can make decisions, such as comparing `pm_state_info` with the `power-state` nodes referenced by the `reinit-power-states` phandles, and call the appropriate reinit function, should `PM_DEVICE_ACTION_RESUME` be the triggering action.

The downside is that this is a major change across the Zephyr codebase. I'll need to update all drivers' implementation of the `*_pm_action` callback, so I'd really like @ceolin's input before adding that to this PR.

**Final note:**
Using the drivers' `_pm_action` callback to restore based on system power state is a much cleaner approach, but it does leave open a corner-case for users enabling `CONFIG_PM` without `CONFIG_PM_DEVICE`. In this case, peripherals won't be restored when the system returns from STOP2, I think we should document this somewhere, so that these users don't end up having to chase this down this PR.

ceolin replied:
I don't necessarily think it is a bad idea passing the system state for the device, but as you mentioned that is a major change across the code base and until now that is the first case where we see this needed. Can't you simply use `pm_state_next_get()` in the driver's `pm_action`  to check which power state triggered it ?

This function returns the state selected by the policy and if you call it from a  device `pm_action` it will give which state the SoC will go.

knthm replied:
After checking the code and seeing that `pm_state_next_get(cpu)` directly returns the passed CPU's `pm_state_info`, I guess that could work. I'd originally written off that function due to it having the word *next* in it, as I'd assumed it would return the state following the current state, e.g. ACTIVE or RUNTIME_IDLE.

However, that would still make correct peripheral functionality dependent on `PM_DEVICE` and `!PM_DEVICE_RUNTIME_EXCLUSIVE`, which is not the the default configuration when enabling simple PM, with our without device PM. I other words it's still lead to broken functionality for the majority of users who don't explicitly disable `PM_RUNTIME_DEVICE_EXCLUSIVE`, or who only use `CONFIG_PM` without device PM enabled.

Could you have a look at my answer to your other comment [below](https://github.com/zephyrproject-rtos/zephyr/pull/59200#discussion_r1245798774).

ceolin replied:
Sorry, not getting your point here. If you don't enable `PM_DEVICE` you don't have device power management support ... in other words no `pm_device_action_run` . No one will trigger device power management.

And if you have `PM_RUNTIME_DEVICE_EXCLUSIVE` enabled, the kernel will not suspend/resume devices you have to it manually.

knthm replied:
Now uses reworked `pm_notifier` logic from #60998.

erwango reviewed (commented):
Thanks for taking a stab at this!

Initial "tires kicking" review. Great work but there are some points I'd like to refine or discuss, specially the integration with PM framework (adding @ceolin to the review for this specific part). Anyway, this goes in the good direction.

knthm commented:
- Uncomment code left over from development & testing
- Rebase on main

knthm commented:
- Add `uart_config` struct to persist runtime application/subsystem initialized UART parameters
- Consolidate UART parameter configuration between driver instance (re-)init and runtime configuration execution flows
- Additional `uart_stm32` refactoring (See [here](https://github.com/zephyrproject-rtos/zephyr/pull/59200#discussion_r1232836255) for more details)

knthm commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -44,6 +46,20 @@ LOG_MODULE_REGISTER(i2c_ll_stm32);
+				power_state_name \
+			) \
+		), \
+		DT_PROP_BY_PHANDLE_IDX(node_id, prop, idx, \
+			substate_id) \
```
#### Note to self

`substate-id` is not a required property, so we should first check if it is set before generating the `pm_func` call here.

knthm commented:
- Fix macro indentation in `uart_stm32.c`
- Move `reinit-power-states` property to `base/pm` DT binding
- Fix corner-case of `pm_policy_state_lock_` call generation macro for when a `power-state`'s `substate-id` isn't set

gmarull commented on dts/bindings/base/pm.yaml:
```diff
@@ -30,3 +30,9 @@ properties:
     description: |
       Automatically configure the device for runtime power management after the
       init function runs.
+
+  reinit-power-states:
+    type: phandles
+    description: |
+      System power states in which a peripheral's configuration is lost, and
+      must be reinitialized to ensure correct operation.
```
this is something that is known by the system, ie, driver, so why put it in DT?

knthm replied:
This is different between every SoC series, e.g. SPI loses context in STOP0 on a STM32WB, but on STM32WL it's not until STOP2.

ceolin replied:
The initial assumption was that devices state would be related with a power state but in practice it is fragile and since substates  addition they are even more de-coupled ... Basically the only relationship between them now is which power state will trigger devices suspensions ... that is reasonable for me.

erwango replied:
If we're fine with this proposal, I'd proposal to move this  and related commits in a dedicated PR

ceolin replied:
Please, do this.

knthm replied:
Please see #60369

gmarull commented on soc/arm/st_stm32/stm32wl/power.c:
```diff
@@ -76,20 +86,34 @@ __weak void pm_state_set(enum pm_state state, uint8_t substate_id)
+		case 2: /* STOP1 */
+			break;
+		case 3: /* STOP2 */
+			/* Reinitialize peripheral functionality */
+			DT_FOREACH_STATUS_OKAY(st_stm32_adc,    STM32_ADC_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_i2c_v2, STM32_I2C_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_spi,    STM32_SPI_REINIT);
+			DT_FOREACH_STATUS_OKAY(st_stm32_usart,  STM32_USART_REINIT);
```
definitely -1 on calling driver code from soc/. we need a mechanism so that drivers can subscribe to system power state changes and act accordingly.

knthm replied:
Yes this is messy, I agree.
I'm currently replacing it with a callback similar to pm_action_cb in the pm_device struct.

ceolin replied:
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/pm/pm.h#L83-L91

PM subsystem send notifications when entering/leaving a power state, you can subscribe to receive them.

knthm replied:
I saw the `pm_notifier_` functions, but that unfortunately only gives high-level PM state information via the `pm_notifier` struct, i.e. `pm_state` enum. We need access to the `substate_id` to determine if we're coming out of e.g. STOP2 where we must reinit the peripheral, or STOP1 were we shouldn't.

My approach right now is still adding a `pm_reinit_cb` to the `pm_device` struct, which will be called via its own function inside `pm.c` `pm_resume_devices` and `pm_suspend_devices`, where `pm_resume_devices` is also called, but independent of `CONFIG_PM_DEVICE` and `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE`.

Looking at the notifier stuff now, I suppose I could also extend the `pm_notifier` struct with `substate_id`, would you prefer this approach to the one above?

Trying to hammer this out in between the EOSS in-person meetings and talks, without losing too much mental context from the first two. :)

ceolin replied:
I think the approach here is the problem ... Are you trying to suspend / resume devices manually because the subsystem is not doing it for  your state ? If that is the case I think the right thing is make this selection of which power state triggers device pm more generic (using device tree as you mentioned). In that case you can query the system state in your `pm_action_cb`. In this case when a device is asked for suspend you can see what is the next state the SoC is going and when it is asked to resume you can see what is the state that the SoC is leaving.

knthm replied:
Sorry, I must not have explained it clearly enough.

So the issue I'm trying to resolve here is in System PM. Issue example:

1. STM32WL SoC goes into STOP2. suspend-to-idle System PM low power mode
2. Some wakeup event wakes the SoC and CPU
3. The I2C register contents and clock. configuration have been lost, due to the way the STOP2 SoC low power mode works. These must now be reinitialized for I2C to function again.

My proposed `pm_reinit_cb` solution above merely piggybacks off of the `pm_device` struct, which is still present, even when PM Device is disabled due to the way the PM Device init macro in the driver works.

Are you still in Prague this Friday? I feel a 5 minute chat would be much more productive than this back and forth, should you have a minute :)

ceolin replied:
> Sorry, I must not have explained it clearly enough.
> 
> So the issue I'm trying to resolve here is in System PM. Issue example:

Don't need to sorry at all :)

> 
> 1. STM32WL SoC goes into STOP2. suspend-to-idle System PM low power mode
> 2. Some wakeup event wakes the SoC and CPU

In this point I see two places where you can hook this code (both requires changes)

1) With the proposed changes to tell the system which states devices should be suspended / resumed, the pm subsystem can call `pm_device_action_run(dev, PM_DEVICE_ACTION_RESUME);` and you can check in the callback you can check if the system is returning from STOP2. This option assumes you have `CONFIG_PM_DEVICE` enabled.

2) Add more notifications ... 
```
pm_state_notify(true); /* entering STOP2 */
pm_state_set(STOP2);
pm_state_notify(false); /* Leaving STOP2 */
pm_state_exit_pos_ops(STOP2);
pm_state_notify(true); /* entering ACTIVE */

```

> 3. The I2C register contents and clock. configuration have been lost, due to the way the STOP2 SoC low power mode works. These must now be reinitialized for I2C to function again.
> 
> My proposed `pm_reinit_cb` solution above merely piggybacks off of the `pm_device` struct, which is still present, even when PM Device is disabled due to the way the PM Device init macro in the driver works.
> 
> Are you still in Prague this Friday? I feel a 5 minute chat would be much more productive than this back and forth, should you have a minute :)

knthm replied:
1. Again, this depends on a very uncommon and explicit configuration combination of `CONFIG_PM_DEVICE` and `!CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE`; not to mention that it's required logic for correct `CONFIG_PM` operation on affected SoCs and handling this with `CONFIG_PM_DEVICE` logic is out-of-scope for that, IMHO.
2. Using the notifier feature, even if modified to additionally execute before `pm_exit_post_ops`, adds runtime overhead for all returning low-power states in all drivers that make use of reinit.
My new PR #68592 will execute only when returning from the affected `power-states` and will be compiler-optimized to noop on platforms/SoCs/board configurations where it's not needed.

gmarull commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -887,34 +887,73 @@ static inline bool spi_stm32_is_subghzspi(const struct device *dev)
+ *
+ * @return 0
+ */
+int spi_stm32_reinit(const struct device *dev)
+{
+	const struct spi_stm32_config *cfg = dev->config;
+	int err;
+
+	if (!device_is_ready(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE))) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	err = spi_stm32_clock_configure(cfg);
+	if (err < 0) {
+		return err;
+	}
+
+	return 0;
+}
+
+static int spi_stm32_init(const struct device *dev)
+{
+	struct spi_stm32_data *data __attribute__((unused)) = dev->data;
+	const struct spi_stm32_config *cfg = dev->config;
+	int err;
+
+	if (!device_is_ready(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE))) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	err = spi_stm32_clock_configure(cfg);
+	if (err < 0) {
+		return err;
+	}
```
this is introducing code duplication. what is really required after stop2? only clock re-enablement or state recovery as well?

knthm replied:
Done.

erwango commented on dts/arm/st/wl/stm32wl.dtsi:
```diff
@@ -258,6 +258,7 @@
 			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00200000>;
 			interrupts = <30 0>, <31 0>;
 			interrupt-names = "event", "error";
+			reinit-power-states = < &stop2 >;
```
Nit:
```suggestion
			reinit-power-states = <&stop2>;

```

erwango commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -127,14 +143,33 @@ i2c_stm32_transaction(const struct device *dev,
 			flagsp = next_msg_flags;
 		}
+
+#ifdef CONFIG_PM
+		/* Prevent SoC from entering power states defined in
+		 * reinit-power-states DT property
+		 */
+		DT_FOREACH_PROP_ELEM_VARGS(
+			DT_DRV_INST(0), reinit_power_states,
+			PM_POLICY_STATE_LOCK_ACTION, pm_policy_state_lock_get);
+#endif
```
Can we think of a generic, dedicated macro ? (ideally with CONFIG_PM=n abstraction)
Also, it should take an argument that would be the "lowest level unless specified otherwise in dt" so that `reinit-power-states` can remain optional.

knthm replied:
I'll attempt to create an intermediate macro for generating this logic directly.

knthm replied:
Please see https://github.com/zephyrproject-rtos/zephyr/pull/60369#issuecomment-1641248231

erwango reviewed (changes requested):
Proposal to split non controversial part into its own PR to move on.

knthm committed:
doc: Add reinit-power-states to System PM docs

Add general purpose of and example use-case of the reinit-power-states
property added in this commit series.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
pm: Replace state transition bool with bit field

Add bit field values for specifying PM state transition direction, e.g.
state entry, exit, or both; use these in place of entering_state bool.

Move exising and add new pm_state_notify calls to the exact edges of
state transitions, to provide more flexibility and cover more use
cases.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
pm: Refactor pm_notifiers to be power-state specific

Replace individual entry and exit callbacks with a singular callback,
with a separate direction bit field to indicate whether the struct
applies to state entry/exit, or both.

Add a void pointer to provide optional context for the callback.

Replace the singular pm_notifiers slist with individual power state-
specific slists, generated from DT power-states definitions on SMP, and
cpu-power-states on single-CPU platforms.

Un/register functions now take a power (sub)state as a parameter, in
addition to the usual pm_notifier struct; used to place the notifier
into the correct list.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
tests: Update tests/samples affected by pm_notifier API change

Update pm tests and sample that make use of pm_notifier functionality,
or test system power state transitions.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
drivers: serial: stm32: Add reinitialization feature

The added reinitialization function helps PM logic restore peripheral
functionality when clock context and register contents are lost in
certain low-power modes.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
pm: state: Add power-states array index helpers

Add pmstates array generated from global power-states DT node children,
vs. current per-CPU only pmstates array (SMP platforms only).

Add helper functions for retrieving the index of a global/per-CPU
power-state array element.
These functions are intended to faciliate easier processing of similar
PM data structures generated from the same DT power-states nodes.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
pm: state: Add helper macros for reinit states

Add helper macros for building a pm_state_info array from a device's
reinit-power-states.

These macros are intended to assist drivers in and around device
reinitializiation.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm committed:
drivers: spi: stm32: Add reinitialization feature

The added reinitialization feature helps PM logic restore bus
functionality when clock context is lost in certain low-power modes.

Signed-off-by: Kenneth J. Miller <ken@miller.ec>

knthm commented on drivers/serial/uart_stm32.c:
```diff
@@ -1855,6 +1879,16 @@ static int uart_stm32_init(const struct device *dev)
+		for (uint8_t i = 0; i < config->reinit_states_size; i++) {
+			pm_notifier_register(config->notifier, config->reinit_states[i].state,
+					     config->reinit_states[i].substate_id);
+			/* TODO: update exit latency here? */
```
Remove this, silly!

erwango commented:
@ceolin How do you feel globally about this PR ?

I'm willing to approve https://github.com/zephyrproject-rtos/zephyr/pull/59200/, but this will be deadcode if https://github.com/zephyrproject-rtos/zephyr/pull/60998 never gets in.

ceolin commented:
> @ceolin How do you feel globally about this PR ?
> 
> I'm willing to approve #59200, but this will be deadcode if #60998 never gets in.

@erwango it seems that the notifier changes is replicated in all prs (#59200, #61680 and #60998). Will focus the review first in #60998 that is base for all of them.
